core/ir/x86/decode.c

```diff
@@ -2579,6 +2580,14 @@ check_is_variable_size(opnd_t op)
 static byte *
 decode_common(dcontext_t *dcontext, byte *pc, byte *orig_pc, instr_t *instr)
 {
+    /* Synthetic ISA has its own decoder.
```

Comment:
What about other decoding routines like decode_sizeof(), etc.?

Reply:
I can implement it, but I need to understand if we need it for the decoding of Synthetic ISA instructions (do we?).
There are several API in core/ir/decode_api.h and core/ir/encode_api.h does Synthetic ISA need to support all of them?
My understanding was that we only needed `decode()` and `instr_encode()`.

Reply:
We have to make sure someone else can understand what is available: having a new ISA implies the whole API works.  How would they know what works?  If they tried decode_sizeof() they wouldn't get an error, they'd get x86 sizes or sthg?  That seems very very confusing.  We have to clearly document which pieces are supported (ideally in both a general section on the new ISA and in each API function where it's not supported?  maybe the former is enough?) and ideally have errors on the unsupported ones.

Reply:
Ah, yes, we should document what's supported and what's not, and ideally we'd return an error if DR_ISA_SYNTHETIC is not supported.
However, since DR_ISA_SYNTHETIC is not a "real" ISA, I don't think we can keep these "return an error" localized to core/ir/synthetic as it currently happens for other architectures.
e.g., in riscv we have a bunch of ASSERT_NOT_IMPLEMENTED localized in core/ir/riscv64
```
byte *
decode_opcode(dcontext_t *dcontext, byte *pc, instr_t *instr)
{
    /* FIXME i#3544: Not implemented */
    ASSERT_NOT_IMPLEMENTED(false);
    return NULL;
}
```
But we can't do the same for synthetic ISA.

So, we'd have to sprinkle a bunch of:
```
if (instr_get_isa_mode(instr) == DR_ISA_SYNTHETIC)
    ASSERT_NOT_IMPLEMENTED(false);
```
and "pollute" every public API we don't need/want to support.
Which is fine if that's what we want to go with, but it looks very ugly IMHO.
Unless I'm missing something that would avoid doing that.

Perhaps documenting that the synthetic ISA only supports decode() and instr_encode() in core/ir/synthetic/encoding_common.h (where the encoding scheme is described) is sufficient?

Reply:
I'm wondering what it will look like with true multi-ISA support: some kind of dispatch for each interface.  But probably it is best to not try to take that step in this PR.  OK, so we have a single exception for now where for decode() (best to check if any other public functions end up calling decode_common()) if the instr has the synthetic ISA set that takes precedent and it ignores the dcontext: on x86 only.  So we document that.

Reply:
Some kind of dispatch based on instr_t isa_mode perhaps.
Documented this in core/ir/encoding_common.h